### ISSUE TITLE
CvtToFuzzyMeanToMid DuplicateRawValues Issue

### ISSUE THREAD

**Issue by zcanter** [consbio/mpilot#12]
CvtToFuzzyMeanToMid throw the DuplicateRawValues error (inside ConvertToFuzzyCurve) when `count(raw_vales.unique()) <= len(fuzzy_values)`

**Example**

```
x = [0, 25, 88, 999]
ConvertToFuzzyMeanToMid(
     input_vals = [x]
     fuzzy_vals = [0, 0.25, 0.5, 0.75, 1]
     ignore_zeros = True/False
)
# Output: DuplicateRawValues Error

```

**Possible Solution:**

```
# line 215 fuzzy.py
low_value = arr.min()
high_value = arr.max()

if ignore_zeros:
    arr = arr[arr != 0]

mean_value = arr.mean()
below_mean = arr[arr <= mean_value]
above_mean = arr[arr > mean_value]

high_mean = above_mean.mean()
low_mean = below_mean.mean()

raw_values = [low_value, low_mean, mean_value, high_mean, high_value]

if len(raw_values) != len(set(raw_values)):
    print("Duplicate Error")
    # run Normalize(low_value, high_value, 0, 1)
else:
    # run CvtToFuzzyCurve(raw_values, kwargs["FuzzyValues"])
```

Feel free to contact me if you have any questions. If you would like me to try submitting working code via GitHub, I can try and do that as well. Thanks, y'all!

**Comment by nikmolnar** [comment]
It looks like this happens because the `high_mean` and `high_value` values end up being the same (999).

Can you explain the reasoning behind your proposed fix?

**Comment by zcanter** [comment]
Sure, if there are insufficient values to compute the MeanToMid algorithm, the simplest option is to Normalize with the low and high values of the fuzzy_values. Also, John suggests we do this, and I agree. This error is part of a model that we have to finish in mid-September for the Santa Barbara Botanic Garden. Thank you!

**Comment by johngallo** [comment]
some more context: we have an EEMS model that we apply to every invasive species in the region.  And then the results are combined, but that is not uimportant here. The EEMS model works fine when the invasive species has a lot of populations.  In those casees mean to mid value is the best way of addressing four of the operators.  But when there are just a few populations, then men-to-Mid crashes (as per above).    ALSO, in the above proposed solution, a warning message should be provided to the user that the particular Mean to Mid command did not have enough data points, and a regular Normalize command was used.

**Comment by mikegough** [comment]
I think that if there aren't enough input values to perform the calculation, it _should_ throw an error, not change the conversion method. The onus should be on the user to ensure that the input data they are providing conforms to the requirements needed to make the calculation. 

What about having two models, one that uses the MeanToMid, and one that uses Normalize?  If a species has at least 5 unique input values, call the former. If not, call the latter. 

I don't know how many species you have, but you could even do this programmatically using a Search Cursor in arcpy: 

```
count_unique_values = len(set(row[0] for row in arcpy.da.SearchCursor(species_dataset, species_field)))

if count_unique_values >= 5:
   # call the model that uses MeanToMid
else:
   # call the model that uses Normalize 
```

**Comment by zcanter** [comment]
Okay thanks, Mike, I will look into that solution today as an immediate fix. 

Also, just FYI the python notebook I use mpilot heavily in is [here](https://github.com/consbio/SBBG-EEMS/blob/main/4_EEMS_Model.ipynb).  Also, there is [this notebook](https://github.com/consbio/SBBG-EEMS/blob/main/MeanToMid_ErrorReport.ipynb) about this error. Thanks again all!

**Comment by nikmolnar** [comment]
@mikegough, based on the `MeanToMid` logic, it looks like this isn't purely a matter of the number of values, but rather particularly large values near one end or another. In fact, 4 values could work, as long as the values were more evenly distributed.

For example,if the input  values were `[0, 25, 50, 75]`, then you'd end up with raw values of `[0, 12.5, 37.5, 62.5, 75]` and would not run afoul of the "duplicate values" error when calling `CvtToFuzzyCurve`.

**Comment by nikmolnar** [comment]
Similarly, I could imagine an input with 5 values that would also fail, if the mean fell between the second-to-last and the last values. E.g., `[0, 25, 50, 75, 999]`. In fact, you could theoretically have _any_ number of values, and still cause this error as long as the last value were large enough in comparison with the others.

**Comment by zcanter** [comment]
@nikmolnar Ah, that is a great observation. That is the true issue. I was seeing the symptom instead of the illness. In practice, the tool succeeds if there are more than 20 or so unique values (i.e. chances become really slim the above scenario happens).

**Comment by mikegough** [comment]
Ah, thanks for the clarification. That makes sense.

I still don't like the idea of changing the conversion method instead of throwing an error, but that is one possible solution.

Here's another possible solution: simply drop the low_mean or the high_mean when these are the same as the low or high value, respectively.

Because unless I'm mistaken, when this situation crops up, there won't be any input values between the low and the mean, or if the outlier is in the other direction, between the mean and the high. And if there aren't any values within these ranges to perform a transformation on, you don't lose anything by dropping the offending low_mean or high_mean duplicate.

For example, given Zac's input values:

[0, 25, 88, 999]

The MeanToMid algorithm would normally generate the following raw values:

low: 0
low_mean: 37.6666666667
mean: 278
high_mean: 999
high: 999

And if you simply dropped the high_mean value (and the corresponding fuzzy value of 0.75), that wouldn't have any effect on the transformation.

Here's another question for Zac and John:

Does the 999 represent NoData? If so, do you really want that value being used in either the MeanToMid or the Normalization transformations?

Because another idea that comes to mind would be to change the "IgnoreZeros" parameter on the MeanToMid tool to "Ignore a list of Values"...

IgnoreValues: [999, -999, etc.]

This would not only fix the problem at hand (assuming 999 represents NoData), but it would also be useful in general for excluding outliers or NoData values.

**Comment by zcanter** [comment]
I think that's a good idea. Regarding your question, no 999 is not NODATA. We use 0 for that and we carefully use the ignore_zeros parameter of MeanToMid

**Comment by nikmolnar** [comment]
This is fixed in version `1.0.0`.
